### ISSUE TITLE
Issues with `Window::request_redraw()` on X11

### ISSUE THREAD

**Issue by JohnDoneth** [rust-windowing/glutin#1193]
## Platform
Platform: Linux with X11. The same code works on Windows.

## Issue

- ~~Calling `Window::request_redraw()` at all while handling `EventsCleared` causes an event stalling issue where events are very delayed.~~
- ~~Calling `Window::request_redraw()` while handling `RedrawRequested` freezes the program.~~ Fixed by [winit PR 1046](https://github.com/rust-windowing/winit/pull/1046)
- Calling `Window::request_redraw()` causes a large delay in event processing.

## Possible Cause

~~I'm unsure if the issues are related but I believe I found the cause to the second issue and I believe it to be a deadlock on a `Mutex`. The implementation of `EventLoop`'s has a `Mutex` field called `pending_redraws` that is locked when dispatching `RedrawRequested` events. Because of this when I try and request a redraw after the frame is over during the `RedrawRequested` event handler, the `Mutex` is still locked and a deadlock occurs when trying to acquire it as `Window::request_redraw` tries to do so.~~

~~https://github.com/rust-windowing/winit/blob/master/src/platform_impl/linux/x11/mod.rs#L281
https://github.com/rust-windowing/winit/blob/master/src/platform_impl/linux/x11/window.rs#L1288~~

**Comment by ZeGentzy** [comment]
Dupe: https://github.com/rust-windowing/glutin/issues/1054 ?

**Comment by ZeGentzy** [comment]
@murarth Can you look into this? This issue is really bugging me.

**Comment by JohnDoneth** [comment]
I possibly found the bug?
Changing
```rust
self.inner_loop
    .dispatch(Some(::std::time::Duration::from_millis(0)), &mut ())
    .unwrap();
```
in `platform_impl/linux/x11/mod.rs` to
```rust
self.inner_loop
    .dispatch(Some(::std::time::Duration::from_millis(1)), &mut ())
    .unwrap();
```
Seems to fix the event loop completely. I think the duration just has to be non-zero for events to be properly dispatched. I was able to get my program working perfectly with just that change.

**Comment by ZeGentzy** [comment]
@chrisduerr Does this fix #1054?
@JohnDoneth Can you figure out where in mio the time is being spent when we use ::std::time::Duration::from_millis(0)?

Afaik, it should time out instantly: https://docs.rs/mio/0.6.19/src/mio/poll.rs.html#1086

**Comment by murarth** [comment]
I've found the problem and submitted a PR with a more reliable solution here: rust-windowing/winit#1048

The problem is that, by sending a message to the X server to flip the screen buffer (and reading for a success/error response), winit is likely to receive user input messages which were pending on the socket. Xlib silently buffers these until you request them with a more general event reading call. In the meantime, though, the socket now has no data waiting; therefore, the event loop will not wake up and process the events buffered in Xlib.

(And adding time to the event loop poll would sometimes work, only because the delay meant more time to receive another event from the X server, putting the socket into a readable state again.)
